cdc/cdc.cc

```diff
@@ -0,0 +1,221 @@
+ */
+
+#include <utility>
+
+#include "cdc/cdc.hh"
+#include "database.hh"
+#include "db/config.hh"
+#include "dht/murmur3_partitioner.hh"
+#include "schema.hh"
+#include "schema_builder.hh"
+#include "service/migration_manager.hh"
+#include "service/storage_proxy.hh"
+#include "service/storage_service.hh"
+
+using locator::snitch_ptr;
+using locator::token_metadata;
+using locator::topology;
+using seastar::sstring;
+using service::migration_manager;
+using service::storage_proxy;
+
+namespace cdc {
+
+sstring log_name(const sstring& table_name) {
+    static constexpr auto cdc_log_suffix = "_scylla_cdc_log";
+    return table_name + cdc_log_suffix;
+}
+
+sstring desc_name(const sstring& table_name) {
+    static constexpr auto cdc_desc_suffix = "_scylla_cdc_desc";
+    return table_name + cdc_desc_suffix;
+}
+
+static future<>
+remove_log(db_context ctx, const sstring& ks_name, const sstring& table_name) {
+    try {
+        return ctx._migration_manager.announce_column_family_drop(
+                ks_name, log_name(table_name), false);
+    } catch (exceptions::configuration_exception& e) {
+        // It's fine if the table does not exist.
+        return make_ready_future<>();
+    }
+}
+
+static future<>
+remove_desc(db_context ctx, const sstring& ks_name, const sstring& table_name) {
+    try {
+        return ctx._migration_manager.announce_column_family_drop(
+                ks_name, desc_name(table_name), false);
+    } catch (exceptions::configuration_exception& e) {
+        // It's fine if the table does not exist.
+        return make_ready_future<>();
+    }
+}
+
+future<>
+remove(db_context ctx, const sstring& ks_name, const sstring& table_name) {
+    return when_all(remove_log(ctx, ks_name, table_name),
+                    remove_desc(ctx, ks_name, table_name)).discard_result();
+}
+
+static future<> setup_log(db_context ctx, const schema& s) {
+    schema_builder b(s.ks_name(), log_name(s.cf_name()));
+    b.set_default_time_to_live(gc_clock::duration{86400}); // 24h
+    b.set_comment(sprint("CDC log for %s.%s", s.ks_name(), s.cf_name()));
+    b.with_column("stream_id", uuid_type, column_kind::partition_key);
+    b.with_column("time", timeuuid_type, column_kind::clustering_key);
+    b.with_column("batch_seq_no", int32_type, column_kind::clustering_key);
+    b.with_column("operation", int32_type);
+    b.with_column("ttl", long_type);
+    auto add_columns = [&] (const auto& columns) {
+        for (const auto& column : columns) {
+            b.with_column("_" + column.name(), column.type);
+        }
+    };
+    add_columns(s.partition_key_columns());
+    add_columns(s.clustering_key_columns());
+    add_columns(s.static_columns());
+    add_columns(s.regular_columns());
+    return ctx._migration_manager.announce_new_column_family(b.build(), false);
+}
+
+static future<> setup_desc(db_context ctx, const schema& s) {
+    schema_builder b(s.ks_name(), desc_name(s.cf_name()));
```

Comment:
What's a desc?

Reply:
description

Reply:
I guessed it's a description, but it's still a vague name. Maybe setup_stream_description_table()?